clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -94,7 +94,7 @@ private Utils() {}
 
     // This matches URIs of formats: host:port and protocol://host:port
     // IPv6 is supported with [ip] pattern
-    private static final Pattern HOST_PORT_PATTERN = Pattern.compile("^(?:[a-zA-Z][a-zA-Z\\d+-.]*://)?\\[?([0-9a-zA-Z\\-._%:]+)\\]?:([0-9]+)$");
+    private static final Pattern HOST_PORT_PATTERN = Pattern.compile("^(?:[a-zA-Z][a-zA-Z\\d+_.-]*://)?\\[?([0-9a-zA-Z\\-._%:]+)]?:([0-9]+)$");
```

Comment:
Reviewing the pattern compared with the original that the previous PR broke, there are some subtle differences. Because the patterns are so hard to read, I think it's good to keep the change as simple as possible.

* Formerly, the majority of matching was done using `0-9a-zA-Z\-%._`. I think it would be a good idea to use this as a starting point for all of the matching, making minimal changes to this as required for the desired effect. This PR is needed because `_` was omitted, which is easy to miss.
* `0-9` was used to match digits in all cases, except now there's a change to `\d`. Not sure why it was necessary to switch.
* The initial character of the protocol name is now required to be alphabetic, where previously any of the permitted characters in the protocol could occupy any position.
* The protocol name was previously able to include %. I'm sure it's unlikely this is a problem, but it's a change.

Reply:
Hi @AndrewJSchofield 
Thanks for your your comment. 😄 
I agree that we should keep regex easy to read instead of figuring out what each character mean. I will do a change.